File: Huawei/数字字符串组合倒序.java


import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class T45 {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		String input = sc.nextLine();
		// I am an 20-years out--standing @ * -stu-dent
		String str1 = "[^0-9a-zA-Z-]";// 非法字符 //空格 @ *
		String[] items1 = input.split(str1);
		List<String> wordList = new ArrayList<String>();
		for (String w : items1) {
			if (w.replaceAll(" ", "").length() == 0) {
				continue;
			}
			if (w.indexOf("--") != -1) {
				w = w.replaceAll("--", "&");// 后面用&符号统一分割一次
			}
			if (w.indexOf("-") != -1) {
				// 解决头尾出现的问题 即可
				if (w.charAt(0) == '-' || w.endsWith("-")) {
					w = w.replaceAll("-", "&").replace(" ", "");
				}
			}
			for (String item : w.split("&")) {
				if (item.replaceAll(" ", "").length() == 0) {
					continue;
				}
				wordList.add(item);
			}
		}
		for (int i = wordList.size() - 1; i >= 0; i--) {
			System.out.print(wordList.get(i) + " ");
		}
	}
}

File: Huawei/数字字符串组合倒序.py


import re
s = input()
# s="I am an 20-years out---standing @*-stu-dent"
def result():
    # 正则表达式，^ 表示匹配字符串的开始位置 (例外 用在中括号中[ ] 时,可以理解为取反,表示不匹配括号中字符串)
    # $ 表示匹配字符串的结束位置
    # 非字母、数字、-的字符都当成单词之间的分隔符
    words = re.split(r"[^0-9a-zA-Z\-]", s)

    print(words)
    ans = []
    for i in range(len(words)-1, -1, -1):
        word = words[i]
        if word == "":
            continue
        word = word.strip("-") # 去除-单词左右边不能是 "-"
        # { } 大括号用于限定匹配次数 如 {n}表示匹配n个字符 {n,}表示至少匹配n个字符 {n,m}表示至少n,最多m
        subword = re.split(r"-{2,}", word) # "--"为分割符
        for j in range(len(subword)-1, -1, -1):
            if subword[j] != "":
                ans.append(subword[j])
    return " ".join(ans)
print(result())